### PR TITLE
Messaging implementation /backend  #3

### PR DESCRIPTION
**Pull request by fetahokey**
### What this PR does :
-  ➕  models [`Conversation` and `Message`]
-  ➕ routes

    | **Function** | **Route**|
    | ------------- | ------------- |
    |  `getAllConversations` |  ` GET /conversations`|
    | `getOneConversation` | `GET /conversations/:id`|
    | `createConversation` |  `POST /conversations`|
    | `createMessage`|  ` POST /messages`|
    | `updateMessageStatus` | `PATCH /messages/:id`|

- ➕ validators 

  | **Conversation**  | **Message** |
  | ------------- | ------------- |
  | `validateCreateConversation  `| `validateCreateMessageRequest `|
  | validateUserProfile  | `validateCreateUpdateMessage`|
  | `validateConversationMemebers`| `validateUpateMessageStatus`|
  | `validateConversationMemebers`| `validateUpdateMessageRequest`|

>this is ready to use postman collection,  you can test all the routes :
> [Messaging.postman_collection.zip](https://github.com/hatchways/team-dumplings/files/7455102/Messaging.postman_collection.zip)

### Information needed to test this feature :
- Configure your DB (.env), 
- Insert some mock-data (`users`, `profiles`)
- Run `npm run dev` and test routes (eg Postman)
- Log in first, Since these routes are private

### Limitations with the current functionality :
- Fetch messages efficiently when we got some big data.

### Features:
- [x] #3 
- [ ] Data security and encryption

### CODE REVIEW

File: server/controllers/message.js [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+  let message = await Message.create({
+    conversationId,
+    sender: profile._id,
+    text,
+  });
+
+  message = await Message.populate(message, {
+    path: "sender",
+    select: { firstName: 1, lastName: 1 },
+  });
```
Comment (ezeilo-su) [inline server/controllers/message.js:33]: Chain these two if possible.

File: server/controllers/message.js [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+  if (message) {
+    res.status(201).json({
+      message,
+    });
+  } else {
+    res.status(500);
+    throw new Error("Internal Server Error");
+  }
+});
```
Comment (ezeilo-su) [inline server/controllers/message.js:43]: :+1:

File: server/controllers/message.js [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+
+// @route PATCH /messages/:id
+// @desc update message `seen` to true, when message is seen
+// @access private
```
Comment (ezeilo-su) [inline server/controllers/message.js:47]: :+1:

File: server/controllers/message.js [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+exports.updateMessageStatus = asyncHandler(async (req, res, next) => {
+  const { conversationId, recipientId } = req.body;
```
Comment (ezeilo-su) [inline server/controllers/message.js:49]: Also, validate so you return a `400` if any of the fields is not supplied by the front-end

File: server/models/Conversation.js [hunk 5]
```diff
@@ -0,0 +1,16 @@
+const mongoose = require("mongoose");
+const Profile = require("./Profile");
+
+const conversationSchema = new mongoose.Schema(
+  {
+    members: [{ type: mongoose.Schema.Types.ObjectId, ref: "Profile" }],
+  },
+  {
+    timestamps: true,
+  }
```
Comment (ezeilo-su) [inline server/models/Conversation.js:10]: :+1:

File: server/models/Conversation.js [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+);
+
+module.exports = Conversation = mongoose.model(
+  "Conversation",
+  conversationSchema
+);
```
Comment (ezeilo-su) [inline server/models/Conversation.js:16]: :+1:

File: server/models/Message.js [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+    text: {
+      type: String,
+      required: true,
+    },
+    sender: {
+      type: mongoose.ObjectId,
+      ref: "Profile",
+      required: true,
+    },
+    seen: {
```
Comment (ezeilo-su) [inline server/models/Message.js:19]: Though, just my opinion, I'd call this `read`

File: server/routes/conversation.js [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+router
+  .route("/:id")
+  .get([protect, validateId, validateUserProfile], getOneConversation);
+
+router
+  .route("/")
+  .post(
+    [protect, validateCreateConversation, validateConversationMemebers],
+    createConversation
+  );
```
Comment (ezeilo-su) [inline server/routes/conversation.js:27]: You can try to be consistent in the way you're passing your middleware. Use an array in all the routes if you prefer that style.

**Review by ezeilo-su — COMMENTED** [review]
Nice job with this one :+1:  Just a few minor comments that I left.
Happy coding!

File: server/models/Message.js (hunk 7 above)
Comment (fetahokey) [reply server/models/Message.js:19]: done ✔️

File: server/controllers/conversation.js [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+  });
+});
+
+// @route POST /conversations
+// @desc create new conversation between authennticated user (sender), and other user (recipient)
+// @access private
+exports.createConversation = asyncHandler(async (req, res, next) => {
+  const { members } = req.body;
+
+  let conversation = await Conversation.findOne({
```
Comment (rajivtitus) [inline server/controllers/conversation.js:52]: 👍🏽

File: server/controllers/conversation.js [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+// @access private
+exports.getAllConversations = asyncHandler(async (req, res, next) => {
+  const userId = req.user.id;
+  const profile = await Profile.findOne({ user: userId });
+
+  const conversations = await Conversation.find({
+    members: { $in: [profile._id] },
+  }).populate("members", "firstName lastName");
+
+  res.status(200).json({ conversations });
```
Comment (rajivtitus) [inline server/controllers/conversation.js:16]: It would been nice to have the last message data sent along with a conversation :)

**Review by rajivtitus — APPROVED** [review]
Great work here! 👍🏽

File: server/controllers/conversation.js (hunk 10 above)
Comment (fetahokey) [reply server/controllers/conversation.js:16]: it's applied in the next [PR ](https://github.com/hatchways/team-dumplings/pull/73)
thank you

**Pull request merged**